### ISSUE TITLE
Can't get edges by id for JanusGraph backend

### ISSUE THREAD

**Issue by mad** [opencypher/cypher-for-gremlin#223]
JanusGraph has bug  https://github.com/JanusGraph/janusgraph/issues/1314

But cypher translation generate complex query for relatively simple operation

Cypher
```
MATCH ()-[r]-() where id(r) = {id} RETURN r
```

Gremlin
```
g.V().as('  cypher.path.start.GENERATED1')
     .bothE().as('r')
     .otherV().as('  UNNAMED14')
     .dedup('  cypher.path.start.GENERATED1', 'r', '  UNNAMED14')
     .where(__.choose(__.constant(2), __.constant(2), __.constant('  cypher.null')).is(neq('  cypher.null')).as('  GENERATED2')
              .select('r')
              .choose(neq('  cypher.null'), __.id(), __.constant('  cypher.null')).is(neq('  cypher.null'))
              .where(eq('  GENERATED2')))
     .select('r')
     .project('r')
     .by(__.choose(neq('  cypher.null'),
                   __.project('  cypher.element', '  cypher.inv', '  cypher.outv').by(__.valueMap(true)).by(__.inV().id()).by(__.outV().id()),
                   __.constant('  cypher.null')))
```
This query doesn't return any result

And profile say:

```
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep(vertex,[])@[  cypher.path.start....                    14          14           6.324    68.16
    \_condition=()
    \_isFitted=false
    \_query=[]
    \_orders=[]
    \_isOrdered=true
  optimization                                                                                 0.017
  optimization                                                                                 0.004
  scan                                                                                         0.000
    \_condition=VERTEX
    \_query=[]
    \_fullscan=true
JanusGraphVertexStep(BOTH,edge)@[r]                                   46          46           0.812     8.75
    \_condition=(EDGE AND visibility:normal)
    \_isFitted=true
    \_vertices=1
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@801a60ee
    \_orders=[]
    \_isOrdered=true
  optimization                                                                                 0.009
  optimization                                                                                 0.003
  optimization                                                                                 0.002
  optimization                                                                                 0.002
  optimization                                                                                 0.002
  optimization                                                                                 0.001
  optimization                                                                                 0.002
  optimization                                                                                 0.002
  optimization                                                                                 0.001
  optimization                                                                                 0.001
  optimization                                                                                 0.003
  optimization                                                                                 0.004
  optimization                                                                                 0.003
  optimization                                                                                 0.003
EdgeOtherVertexStep@[  UNNAMED14]                                     46          46           0.112     1.21
DedupGlobalStep([r,   UNNAMED14,   cypher.path....                    46          46           0.132     1.43
TraversalFilterStep([ChooseStep([ConstantStep(2...                                             1.765    19.03
  ChooseStep([ConstantStep(2), ProfileStep, Has...                    46          46           0.522
    ConstantStep(2)                                                   46          46           0.034
    HasNextStep                                                       46          46           0.046
    ConstantStep(  cypher.null)                                                                0.049
    EndStep                                                                                    0.030
    ConstantStep(2)                                                   46          46           0.041
    EndStep                                                           46          46           0.045
  IsStep(neq(  cypher.null))@[  GENERATED2]                           46          46           0.125
  SelectOneStep(last,r)                                               46          46           0.119
  ChooseStep([LambdaFilterStep(neq(  cypher.nul...                    46          46           0.477
    LambdaFilterStep(neq(  cypher.null))                              46          46           0.043
    HasNextStep                                                       46          46           0.039
    ConstantStep(  cypher.null)                                                                0.041
    EndStep                                                                                    0.028
    IdStep                                                            46          46           0.063
    EndStep                                                           46          46           0.043
  IsStep(neq(  cypher.null))                                          46          46           0.091
  WherePredicateStep(eq(  GENERATED2))                                                         0.087
SelectOneStep(last,r)                                                                          0.126     1.37
ProjectStep([r],[[ChooseStep([LambdaFilterStep(...                                             0.004     0.05
                                            >TOTAL                     -           -           9.278        -
```

On janusgraph this query iterate all vertices

But query like that doesn't affect for 1314 bug and have good explain plan

```
g.E(id).next()
```

**Comment by dwitry** [comment]
Hello @mad,

current Cypher for Gremlin priority is translation completeness. Some optimizing rewriters [are implemented](translation/src/main/scala/org/opencypher/gremlin/translation/ir/rewrite), but there's room for improvement especially for specific Gremlin implementations.

Query complexity can be explained:

1. In current translation, traversal always start with vertex step.
2. [Rewriter](https://github.com/opencypher/cypher-for-gremlin/blob/master/translation/src/main/scala/org/opencypher/gremlin/translation/ir/rewrite/GroupStepFilters.scala) that could simplify this translation does not triggered in this particular case.

To address this issue, `GroupStepFilters` rewriter should be improved.

**Comment by dwitry** [comment]
Closed this in #227
